File: backend/src/models/ThuMoiNhanViec.js
const mongoose = require('mongoose');

const thuMoiNhanViecSchema = new mongoose.Schema({
  ungVienId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'UngVien',
    required: true
  },
  jobId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'TinTuyenDung',
    required: true
  },
  maOffer: {
    type: String,
    unique: true,
    required: true
  },
  viTri: {
    type: String,
    required: true,
    trim: true
  },
  phongBan: {
    type: String,
    trim: true
  },
  mucLuong: {
    coban: {
      type: Number,
      required: true,
      min: 0
    },
    phucap: {
      type: Number,
      default: 0
    },
    thuong: {
      type: String,
      trim: true
    },
    donVi: {
      type: String,
      default: 'VND'
    }
  },
  ngayBatDau: {
    type: Date,
    required: [true, 'Vui lòng chọn ngày bắt đầu làm việc']
  },
  thoiGianThuViec: {
    soThang: {
      type: Number,
      default: 2,
      min: 0
    },
    mucLuongThuViec: {
      type: Number,
      default: 85,
      min: 0,
      max: 100
    }
  },
  dieuKhoan: [{
    type: String,
    trim: true
  }],
  quyenLoi: [{
    type: String,
    trim: true
  }],
  diaDiemLamViec: {
    type: String,
    required: true,
    trim: true
  },
  nguoiLienHe: {
    hoTen: String,
    chucVu: String,
    email: String,
    soDienThoai: String
  },
  hanTraLoi: {
    type: Date,
    required: [true, 'Vui lòng chọn hạn trả lời']
  },
  trangThai: {
    type: String,
    enum: ['Cho_tra_loi', 'Da_chap_nhan', 'Da_tu_choi', 'Het_han', 'Da_huy'],
    default: 'Cho_tra_loi'
  },
  pdfUrl: {
    type: String
  },
  pdfFileName: {
    type: String
  },
  taoBoi: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NguoiDung',
    required: true
  },
  ngayGui: {
    type: Date,
    default: null
  },
  emailDaGui: {
    type: Boolean,
    default: false
  },
  ngayTraLoi: {
    type: Date
  },
  lyDoTuChoi: {
    type: String,
    trim: true
  },
  ghiChu: {
    type: String,
    trim: true
  }
}, {
  timestamps: true,
  collection: 'thu_moi_nhan_viec'
});

// Index
thuMoiNhanViecSchema.index({ ungVienId: 1 });
thuMoiNhanViecSchema.index({ jobId: 1 });
thuMoiNhanViecSchema.index({ maOffer: 1 });
thuMoiNhanViecSchema.index({ trangThai: 1 });
thuMoiNhanViecSchema.index({ ngayGui: -1 });

// Virtual để kiểm tra hết hạn
thuMoiNhanViecSchema.virtual('hetHan').get(function() {
  return new Date() > new Date(this.hanTraLoi);
});

// Tự động tạo mã offer
thuMoiNhanViecSchema.pre('save', async function(next) {
  if (!this.maOffer) {
    const { generateCode } = require('../utils/helpers');
    let maOffer;
    let exists = true;
    
    while (exists) {
      maOffer = generateCode('OFF', 8);
      exists = await this.constructor.findOne({ maOffer });
    }
    
    this.maOffer = maOffer;
  }
  next();
});

// Method chấp nhận offer
thuMoiNhanViecSchema.methods.chapNhan = function() {
  this.trangThai = 'Da_chap_nhan';
  this.ngayTraLoi = new Date();
};

// Method từ chối offer
thuMoiNhanViecSchema.methods.tuChoi = function(lyDo) {
  this.trangThai = 'Da_tu_choi';
  this.ngayTraLoi = new Date();
  this.lyDoTuChoi = lyDo;
};

const ThuMoiNhanViec = mongoose.model('ThuMoiNhanViec', thuMoiNhanViecSchema);

module.exports = ThuMoiNhanViec;

File: frontend/lib/models/thu_moi_nhan_viec.dart
class MucLuongOffer {
  final int coban;
  final int phucap;
  final String? thuong;
  final String donVi;

  MucLuongOffer({
    required this.coban,
    this.phucap = 0,
    this.thuong,
    this.donVi = 'VND',
  });

  factory MucLuongOffer.fromJson(Map<String, dynamic> json) {
    return MucLuongOffer(
      coban: json['coban'] ?? 0,
      phucap: json['phucap'] ?? 0,
      thuong: json['thuong'],
      donVi: json['donVi'] ?? 'VND',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'coban': coban,
      'phucap': phucap,
      'thuong': thuong,
      'donVi': donVi,
    };
  }

  int get tongLuong => coban + phucap;
}

class ThoiGianThuViec {
  final int soThang;
  final int mucLuongThuViec;

  ThoiGianThuViec({
    required this.soThang,
    required this.mucLuongThuViec,
  });

  factory ThoiGianThuViec.fromJson(Map<String, dynamic> json) {
    return ThoiGianThuViec(
      soThang: json['soThang'] ?? 2,
      mucLuongThuViec: json['mucLuongThuViec'] ?? 85,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'soThang': soThang,
      'mucLuongThuViec': mucLuongThuViec,
    };
  }
}

class NguoiLienHe {
  final String? hoTen;
  final String? chucVu;
  final String? email;
  final String? soDienThoai;

  NguoiLienHe({
    this.hoTen,
    this.chucVu,
    this.email,
    this.soDienThoai,
  });

  factory NguoiLienHe.fromJson(Map<String, dynamic> json) {
    return NguoiLienHe(
      hoTen: json['hoTen'],
      chucVu: json['chucVu'],
      email: json['email'],
      soDienThoai: json['soDienThoai'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'hoTen': hoTen,
      'chucVu': chucVu,
      'email': email,
      'soDienThoai': soDienThoai,
    };
  }
}

class ThuMoiNhanViec {
  final String id;
  final String ungVienId;
  final String jobId;
  final String maOffer;
  final String viTri;
  final String? phongBan;
  final MucLuongOffer mucLuong;
  final DateTime ngayBatDau;
  final ThoiGianThuViec thoiGianThuViec;
  final List<String> dieuKhoan;
  final List<String> quyenLoi;
  final String diaDiemLamViec;
  final NguoiLienHe nguoiLienHe;
  final DateTime hanTraLoi;
  final String trangThai;
  final String? pdfUrl;
  final String? pdfFileName;
  final String taoBoi;
  final DateTime? ngayGui;
  final bool emailDaGui;
  final DateTime? ngayTraLoi;
  final String? lyDoTuChoi;
  final String? ghiChu;
  final DateTime createdAt;
  final DateTime updatedAt;

  ThuMoiNhanViec({
    required this.id,
    required this.ungVienId,
    required this.jobId,
    required this.maOffer,
    required this.viTri,
    this.phongBan,
    required this.mucLuong,
    required this.ngayBatDau,
    required this.thoiGianThuViec,
    required this.dieuKhoan,
    required this.quyenLoi,
    required this.diaDiemLamViec,
    required this.nguoiLienHe,
    required this.hanTraLoi,
    required this.trangThai,
    this.pdfUrl,
    this.pdfFileName,
    required this.taoBoi,
    this.ngayGui,
    this.emailDaGui = false,
    this.ngayTraLoi,
    this.lyDoTuChoi,
    this.ghiChu,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ThuMoiNhanViec.fromJson(Map<String, dynamic> json) {
    return ThuMoiNhanViec(
      id: json['_id'] ?? '',
      ungVienId: json['ungVienId'] is String
          ? json['ungVienId']
          : json['ungVienId']?['_id'] ?? '',
      jobId: json['jobId'] is String
          ? json['jobId']
          : json['jobId']?['_id'] ?? '',
      maOffer: json['maOffer'] ?? '',
      viTri: json['viTri'] ?? '',
      phongBan: json['phongBan'],
      mucLuong: MucLuongOffer.fromJson(json['mucLuong'] ?? {}),
      ngayBatDau: DateTime.parse(json['ngayBatDau']),
      thoiGianThuViec: ThoiGianThuViec.fromJson(json['thoiGianThuViec'] ?? {}),
      dieuKhoan: List<String>.from(json['dieuKhoan'] ?? []),
      quyenLoi: List<String>.from(json['quyenLoi'] ?? []),
      diaDiemLamViec: json['diaDiemLamViec'] ?? '',
      nguoiLienHe: NguoiLienHe.fromJson(json['nguoiLienHe'] ?? {}),
      hanTraLoi: DateTime.parse(json['hanTraLoi']),
      trangThai: json['trangThai'] ?? 'Cho_tra_loi',
      pdfUrl: json['pdfUrl'],
      pdfFileName: json['pdfFileName'],
      taoBoi: json['taoBoi'] ?? '',
      ngayGui: json['ngayGui'] != null ? DateTime.parse(json['ngayGui']) : null,
      emailDaGui: json['emailDaGui'] ?? false,
      ngayTraLoi: json['ngayTraLoi'] != null ? DateTime.parse(json['ngayTraLoi']) : null,
      lyDoTuChoi: json['lyDoTuChoi'],
      ghiChu: json['ghiChu'],
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'ungVienId': ungVienId,
      'jobId': jobId,
      'maOffer': maOffer,
      'viTri': viTri,
      'phongBan': phongBan,
      'mucLuong': mucLuong.toJson(),
      'ngayBatDau': ngayBatDau.toIso8601String(),
      'thoiGianThuViec': thoiGianThuViec.toJson(),
      'dieuKhoan': dieuKhoan,
      'quyenLoi': quyenLoi,
      'diaDiemLamViec': diaDiemLamViec,
      'nguoiLienHe': nguoiLienHe.toJson(),
      'hanTraLoi': hanTraLoi.toIso8601String(),
      'trangThai': trangThai,
      'pdfUrl': pdfUrl,
      'pdfFileName': pdfFileName,
      'taoBoi': taoBoi,
      'ngayGui': ngayGui?.toIso8601String(),
      'emailDaGui': emailDaGui,
      'ngayTraLoi': ngayTraLoi?.toIso8601String(),
      'lyDoTuChoi': lyDoTuChoi,
      'ghiChu': ghiChu,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  bool get hetHan => DateTime.now().isAfter(hanTraLoi);
  bool get choTraLoi => trangThai == 'Cho_tra_loi';
  bool get daChapNhan => trangThai == 'Da_chap_nhan';

  String get trangThaiText {
    const map = {
      'Cho_tra_loi': 'Chờ trả lời',
      'Da_chap_nhan': 'Đã chấp nhận',
      'Da_tu_choi': 'Đã từ chối',
      'Het_han': 'Hết hạn',
      'Da_huy': 'Đã hủy',
    };
    return map[trangThai] ?? trangThai;
  }
}